src/AI.cpp

```diff
@@ -0,0 +1,137 @@
+//
+// Created by Carolin on 16.05.2020.
+//
+
+#include "AI.hpp"
+
+#include <generate/ItemChoice.hpp>
+#include <generate/EquipmentChoice.hpp>
+#include <generate/GameOperation.hpp>
+#include "util/Logging.hpp"
+#include "AICallback.hpp"
+
+AI::AI(const std::string &address, uint16_t port, const std::string &name, unsigned int verbosity,
+       unsigned int difficulty,
+       std::map<std::string, std::string> additionalOptions) : address(address), port(port), name(name),
+                                                               difficulty(difficulty), libClientHandler{std::dynamic_pointer_cast<libclient::Callback>(std::make_shared<AICallback>(*this))} {
```

Comment:
muss dieser handler außerhalb der klasse sein? fände es intuitiver wenn die AI klasse direkt die callbacks handelt, wenn das von LibClient her geht...

Reply:
Wollte das eigentlich so machen wie du das vorschlägst.
Allerdings wird bei `AI` der copy Konstruktur implizit entfernt (da es ein LibClient Objekt als member hat und das keinen copy Konstruktor hat) und dann kann ich `this`, also AI, nicht über `std::make_shared` zu einem shared pointer casten, um damit `libClientHandler` zu initialisieren.
Wenn du eine Lösung dafür hast, ändere ich das gerne.

Reply:
Hier muss noch mal gründlich über die Klassenhierarchie nachgedacht werden. Momentan enthält die AI das LibClient objekt, und das LibClient (über den callback, oder direkt) die AI, was zu einer zyklischen dependency führt...

Reply:
Wie ist das denn im Client gemacht? @Marco5de

Reply:
Im Client hat der Controller einen LibClientHandler (ähnlich zu hier) aber der Callback kennt den Controller nicht, weil der Callback einfach events published

Reply:
Denke es ist nicht sinnvoll in der KI ein extra Observer-Pattern dafür zu implementieren.
Da die Callback Methoden für die Verwendung im Client keine Parameter übergeben dürfen, muss in der Implementierung der Callback Methoden irgendwie auf ein LibClient Objekt zugegriffen werden können und dieses LibClient Objekt muss halt auch das Callback kennen.

Die einzige Alternative die mir einfällt ist, dass die KI kein LibClient Objekt, sondern ein Model erstellt und mit diesem und dem KICallback dann ein Network. Das KICallback hätte dann im Konstruktor als Parameter das Model. Allerdings könnte dann die KI Daten im Model setzen, was sie eigentlich nicht tun sollte.
Jetzt ist halt die Frage was das kleinere von beiden Übeln ist?

Reply:
fix idea: LibClient weak_pointer in constructor